astropy/coordinates/velocities.py

```diff
@@ -0,0 +1,74 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Tools for computing velocities and velocity corrections using Astropy
+coordinates and related machinery.
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+from .. import units as u
+
+from .solar_system import get_body_barycentric_posvel
+from .matrix_utilities import matrix_product
+from .representation import CartesianRepresentation, UnitSphericalRepresentation
+from .builtin_frames import GCRS
+
+
+__all__ = ['helio_corr', 'helio_vector']
+
+KPS = u.km/u.s
+
+def helio_vector(t, loc):
+    """
+    Compute the heliocentric velocity correction at a given time and place.
+
+    Paramters
+    ---------
+    t : astropy.time.Time
+        Time of the observation. Can be a Time array.
+    loc : astropy.coordinates.EarthLocation
+        The observer location at which to compute the correction.
+    """
+    vsun = get_body_barycentric_posvel('sun', t)[1]
+    vearth = get_body_barycentric_posvel('earth', t)[1]
+
+    vsunearth = vearth - vsun
+
+    gcrs_p, gcrs_v = loc.get_gcrs_posvel(t)
+
+    vsuntargxyz = vsunearth.xyz + gcrs_v.xyz
+    return CartesianRepresentation(vsuntargxyz.to(KPS))
+
+
+def helio_corr(t, loc, target):
+    """
+    Compute the correction required to convert a radial velocity at a given
+    time and place to a heliocentric velocity.
+
+    Paramters
+    ---------
+    t : astropy.time.Time
+        Time of the observation. Can be a Time array.
+    loc : astropy.coordinates.EarthLocation
+        The observer location at which to compute the correction.
+    target : astropy.coordinates.SkyCoord
+        The on-sky location at which to compute the correction.
+
+    Returns
+    -------
+    vcorr : astropy.units.Quantity with velocity units
+        The heliocentric correction with a positive sign.  I.e., *add* this
+        to an observed radial velocity to get the heliocentric velocity.
+    """
+    vsuntarg_cartrepr = helio_vector(t, loc)
+    gcrs_p, _ = loc.get_gcrs_posvel(t)
+
+    gtarg = target.transform_to(GCRS(obstime=t, obsgeoloc=gcrs_p))
+    targxyz = gtarg.represent_as(UnitSphericalRepresentation).to_cartesian().xyz
+
+    res = matrix_product(vsuntarg_cartrepr.xyz, targxyz)
```

Comment:
OK, real data beats everything! I also think the route is also clearer than the original - it still avoids the expensive `.xyz`. Note that you can remove the `.to_cartesian()` -- `.dot` will do that for you.

But of course need to understand: `represent_as(USph)` is fast - just a new class initialization (this can become faster still...). Then, `to_cartesian` is slightly faster than it would have been before, since no multiplication with a distance (of 1.). While for my original suggestion there is the division by the distance (quite fast, but needless), a similar reinitialization, and then a slightly more expensive conversion to cartesian. My timings are not as different, but still the route via unitspherical is clearly better:
```
g = SphericalRepresentation(np.arange(0,100)*u.deg, np.arange(-45., 55.)*u.deg, np.arange(100,200)*u.kpc)
v = CartesianRepresentation(5., 5., 5., unit=u.km/u.s)
%timeit (g/g.norm()).dot(v)
# 1000 loops, best of 3: 1.69 ms per loop
%timeit g.represent_as(UnitSphericalRepresentation).dot(v)
# 1000 loops, best of 3: 1.49 ms per loop
%timeit g.represent_as(UnitSphericalRepresentation).to_cartesian().dot(v)
# 1000 loops, best of 3: 1.49 ms per loop
```